drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+	int ret;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	priv->vid = vid;
+	ret = rte_vhost_get_negotiated_features(vid, &features);
+	if (ret) {
+		DRV_LOG(ERR, "Failed to get negotiated features.");
+		return ret;
+	}
+	if (RTE_VHOST_NEED_LOG(features)) {
+		ret = rte_vhost_get_log_base(vid, &log_base, &log_size);
+		if (ret) {
+			DRV_LOG(ERR, "Failed to get log base.");
+			return ret;
+		}
+		//TO_DO: add log op
+	}
+
+	//TO_DO: check why ---
+	features |=(1ULL << VIRTIO_F_IOMMU_PLATFORM);
+	priv->guest_features = virtio_pci_dev_features_set(priv->vpdev, features);
+	DRV_LOG(INFO, "vDPA device %d hw feature is %" PRIx64 "guest feature is %" PRIx64,
+					vid,priv->guest_features,features);
+
+	return 0;
+}
+
+static int
+virtio_vdpa_dev_close(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	int ret = 0, i;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+
+	/*suspend*/
+	/*set_vring_base*/
+
+	ret = virtio_pci_dev_interrupts_disable(priv->vpdev, 0);
+	if (ret) {
+		DRV_LOG(ERR, "Error disabling virtio dev interrupts: %s",
+				strerror(errno));
+		return ret;
+	}
+
+	ret = virtio_vdpa_dma_unmap(priv);
+	if(ret) {
+		DRV_LOG(ERR, "Fail to do dma map: %d",ret);
+	}
+
+	virtio_pci_dev_reset(priv->vpdev);
+
+	/* disable all queue*/
+	for (i = 0; i < priv->nr_virtqs; i++) {
+		if (priv->vrings[i]->enable)
+			virtio_vdpa_vring_state_set(vid, i, 0);
+	}
+
+	/* Tell the host we've noticed this device. */
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_ACK);
+
+	/* Tell the host we've known how to drive the device. */
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
+
+	priv->configured = 0;
+
+	/* The mutex may stay locked after event thread cancel - initiate it. */
+	pthread_mutex_init(&priv->vq_config_lock, NULL);
+	DRV_LOG(INFO, "vDPA device %d was closed.", vid);
+	return ret;
+}
+
+static int
+virtio_vdpa_dev_config(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	int ret, fd;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	if (priv->configured) {
+		DRV_LOG(ERR, "Failed to reconfigure vid %d.", vid);
+		return -EINVAL;
+	}
+
+	priv->nr_virtqs = rte_vhost_get_vring_num(vid);
+
+	priv->vid = vid;
+	ret = virtio_vdpa_dma_map(priv);
+	if(ret) {
+		DRV_LOG(ERR, "Fail to do dma map: %d",ret);
+		return ret;
+	}
+	priv->configured = 1;
+	DRV_LOG(INFO, "vDPA device %d was configured.", vid);
+
+	fd =rte_intr_fd_get(priv->pdev->intr_handle);
+
+	if (priv->nvec <= (priv->nr_virtqs +1)) {
+		DRV_LOG(ERR, "Error dev interrupts %d less than queue: %d",
+			priv->nvec, priv->nr_virtqs +1);
+		return -EINVAL;
```

Comment:
you left the device in half configured state with configured = 1 on line 591, and dma map already created,

you need to unwind the error by having jump label goto to do dma_unmap()